src/inmanta/util.py

```diff
@@ -453,3 +454,45 @@ def dfs(node: str, seen: Set[str] = set()) -> None:
         dfs(nodes.pop(0))
 
     return out
+
+
+class _NamedSubLock:
+    def __init__(self, parent: "NamedLock", name: str) -> None:
+        self.parent = parent
+        self.name = name
+
+    async def __aenter__(self) -> None:
+        await self.parent.acquire(self.name)
+
+    async def __aexit__(
+        self, exc_type: Optional[Type[BaseException]], exc_value: Optional[BaseException], traceback: Optional[TracebackType]
+    ) -> None:
+        await self.parent.release(self.name)
+
+
+class NamedLock:
+    """Create fine grained locks"""
+
+    def __init__(self) -> None:
+        self._master_lock: Lock = Lock()
+        self._named_locks: Dict[str, Lock] = {}
+
+    def get(self, name: str) -> _NamedSubLock:
+        return _NamedSubLock(self, name)
+
+    async def acquire(self, name: str) -> None:
+        async with self._master_lock:
+            if name in self._named_locks:
+                lock = self._named_locks[name]
+            else:
+                lock = Lock()
+                self._named_locks[name] = lock
+        await lock.acquire()
+
+    async def release(self, name: str) -> None:
+        async with self._master_lock:
+            lock = self._named_locks[name]
+            lock.release()
+            # This relies on the internal mechanics of the lock
+            if not lock._waiters:
```

Comment:
This is a type error and breaking into an SDK class.

I could not think of a clean alternative.

Suggestions welcome